### ISSUE TITLE
badarg error in binary_to_term/1

### ISSUE THREAD

**Issue by slfritchie** [basho/riak_kv#344]
Riak KV handoff can be stymied indefinitely by a badarg error in erlang:binary_to_term/1, e.g.:

```
2012-06-12 01:25:10.263 [error] <0.89.0> Supervisor riak_core_vnode_sup had child undefined started with {riak_core_vnode,start_link,undefined} at <0.9312.1> exit with reason bad argument in call to erlang:binary_to_term(<<131,104,7,100,0,8,114,95,111,98,106,101,99,116,109,0,0,0,10,114,101,97,100,95,105,110,100,101,...>>) in riak_kv_vnode:do_diffobj_put/3 in context child_terminated
```

Similarly, GET operations can fail with a similar error:

```
2012-06-12 17:47:00.750 [error] <0.89.0> Supervisor riak_core_vnode_sup had child undefined started with {riak_core_vnode,start_link,undefined} at <0.27729.58> exit with reason bad argument in call to erlang:binary_to_term(<<131,104,7,100,0,8,114,95,111,98,106,101,99,116,109,0,0,0,10,114,101,97,100,95,105,110,100,101,...>>) in riak_kv_vnode:do_get_term/3 in context child_terminated
```

In both of these cases, the KV backend is leveldb.  Neither the LevelDB sanity checks (if the bad SST is copied to my local box for testing) nor the Snappy compression library are detecting any error, which is more than a little strange and needs further investigaion.

**Comment by freerobby** [comment]
Not sure if it's the same issue, but I've seen crashes like this, also using leveldb:

```
2013-01-02 04:10:05.371 [error] <0.143.0> Supervisor riak_core_sup had child riak_core_ring_manager started with riak_core_ring_manager
:start_link() at undefined exit with reason bad argument in call to erlang:binary_to_term(<<>>) in riak_core_ring_manager:read_ringfile
/1 line 154 in context start_error
2013-01-02 04:10:15.271 [info] <0.7.0> Application lager started on node 'riak@10.123.21.233'
2013-01-02 04:10:15.345 [error] <0.174.0> CRASH REPORT Process <0.174.0> with 0 neighbours exited with reason: bad argument in call to
erlang:binary_to_term(<<>>) in riak_core_ring_manager:read_ringfile/1 line 154 in gen_server2:init_it/6 line 384
2013-01-02 04:10:25.716 [info] <0.7.0> Application lager started on node 'riak@10.123.21.233'
2013-01-02 04:10:25.835 [error] <0.153.0> CRASH REPORT Process <0.153.0> with 0 neighbours exited with reason: bad argument in call to
erlang:binary_to_term(<<>>) in riak_core_ring_manager:read_ringfile/1 line 154 in gen_server2:init_it/6 line 384
2013-01-02 04:10:35.396 [info] <0.7.0> Application lager started on node 'riak@10.123.21.233'
2013-01-02 04:10:35.579 [error] <0.155.0> CRASH REPORT Process <0.155.0> with 0 neighbours exited with reason: bad argument in call to
erlang:binary_to_term(<<>>) in riak_core_ring_manager:read_ringfile/1 line 154 in gen_server2:init_it/6 line 384
2013-01-02 04:11:50.160 [info] <0.7.0> Application lager started on node 'riak@10.123.21.233'
2013-01-02 04:11:50.249 [error] <0.154.0> CRASH REPORT Process <0.154.0> with 0 neighbours exited with reason: bad argument in call to
erlang:binary_to_term(<<>>) in riak_core_ring_manager:read_ringfile/1 line 154 in gen_server2:init_it/6 line 384
2013-01-02 04:12:28.046 [info] <0.7.0> Application lager started on node 'riak@10.123.21.233'
2013-01-02 04:12:28.166 [error] <0.151.0> CRASH REPORT Process <0.151.0> with 0 neighbours exited with reason: bad argument in call to
erlang:binary_to_term(<<>>) in riak_core_ring_manager:read_ringfile/1 line 154 in gen_server2:init_it/6 line 384
2013-01-02 04:14:03.945 [info] <0.7.0> Application lager started on node 'riak@10.123.21.233'
2013-01-02 04:14:04.029 [error] <0.151.0> CRASH REPORT Process <0.151.0> with 0 neighbours exited with reason: bad argument in call to
erlang:binary_to_term(<<>>) in riak_core_ring_manager:read_ringfile/1 line 154 in gen_server2:init_it/6 line 384
```

**Comment by freerobby** [comment]
After a bit more digging I believe my issue was different, but I'm not positive so I'll post the resolution in case it's handy: somehow my local ring metadata got corrupted. The documentation/fix is below (notice the 0-byte ring file). After deleting it, things were fine.

```
ubuntu@ip-10-123-21-233:/mnt/riak/ring$ ls -alF
total 128
drwxr-xr-x 2 riak riak  4096 Jan  2 02:59 ./
drwxr-xr-x 7 riak riak  4096 Jan  2 04:20 ../
-rw-rw-r-- 1 riak riak 23018 Dec 31 23:34 riak_core_ring.default.20121231233413
-rw-rw-r-- 1 riak riak 22533 Jan  1 23:59 riak_core_ring.default.20130101235914
-rw-rw-r-- 1 riak riak 22565 Jan  2 01:50 riak_core_ring.default.20130102015035
-rw-rw-r-- 1 riak riak 22565 Jan  2 02:56 riak_core_ring.default.20130102025657
-rw-rw-r-- 1 riak riak 22565 Jan  2 02:57 riak_core_ring.default.20130102025736
-rw-rw-r-- 1 riak riak     0 Jan  2 02:59 riak_core_ring.default.20130102025937
ubuntu@ip-10-123-21-233:/mnt/riak/ring$ date
Wed Jan  2 04:35:37 UTC 2013
ubuntu@ip-10-123-21-233:/mnt/riak/ring$ sudo rm riak_core_ring.default.20130102025937
```

**Comment by jtuple** [comment]
As far as I know, this has been fixed in Riak 1.3/1.4 after various changes to LevelDB to make it detect backend errors, plus changes in Riak 1.4 to catch `binary_to_term` errors in general.
